File: 0000-0999/0312.BurstBalloons.go
func maxCoins(nums []int) int {
    // O(n^3)
    nums = append([]int{1}, nums...)
    nums = append(nums, 1)
    n := len(nums)
    dp := make([][]int, n)
    for i := range dp {
        dp[i] = make([]int, n)
    }
    for k := 2; k < n; k++ {
        for left := 0; left < n-k; left++ {
            right := left + k
            for i := left+1; i < right; i++ {
                dp[left][right] = max(dp[left][right], dp[left][i] + nums[left]*nums[i]*nums[right] + dp[i][right])
            }
        }
    }
    return dp[0][n-1]
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: 0000-0999/0312.BurstBalloons.py
class Solution(object):
    def maxCoins(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums = [1] + nums + [1]
        n = len(nums)
        dp = [[0]*n for i in xrange(n)]
        for k in xrange(2, n):
            for left in xrange(0, n-k):
                right = left + k
                for i in xrange(left+1, right):
                    dp[left][right] = max(dp[left][right], nums[left]*nums[i]*nums[right]+dp[left][i]+dp[i][right])
        return dp[0][n-1]
